File: 2024/python/p06.py
# Advent of Code 2024 - Day 6

import tqdm

with open("../input/06.txt") as f:
    raw_data = f.read()

raw_test_data = """....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#..."""


def process(s: str):
    walls = set()
    board = set()
    start = None
    for y, line in enumerate(s.splitlines()):
        for x, c in enumerate(line):
            loc = x + y * 1j
            if c == "#":
                walls.add(loc)
            elif c == "^":
                start = loc
            board.add(loc)
    return start, walls, board


test_data = process(raw_test_data)
data = process(raw_data)


def collect_seen(data):
    start, walls, board = data
    direction = -1j
    seen = set()
    loc = start

    while loc in board:
        seen.add(loc)
        new_loc = loc + direction
        if new_loc in walls:
            direction *= 1j
        else:
            loc = new_loc

    return seen


def part1(data):
    return len(collect_seen(data))


assert part1(test_data) == 41
ans1 = part1(data)
assert ans1 == 5331

## Part 2


def detect_loop(data, new_pos):
    start, walls, board = data
    direction = -1j
    loc = start
    walls = walls.copy()
    walls.add(new_pos)

    states = set()

    while loc in board:
        state = (loc, direction)
        if state in states:
            return True
        states.add(state)

        new_loc = loc + direction
        if new_loc in walls:
            direction *= 1j
        else:
            loc = new_loc

    return False


def part2(data):
    start, walls, board = data
    seen = collect_seen(data)

    places = 0
    for loc in tqdm.tqdm(seen):
        if loc != start:
            if detect_loop(data, loc):
                places += 1

    return places


assert part2(test_data) == 6
ans2 = part2(data)
assert ans2 == 1812

if __name__ == "__main__":
    print(f"Answer 1: {ans1}")
    print(f"Answer 2: {ans2}")

File: 2024/clojure/p06.clj
;; # 🎄 Advent of Code - 2024 - Day 6

(ns p06
  (:require [clojure.string :as str]
            [clojure.test :as test]))

(def data-string (slurp "../input/06.txt"))
(def test-data-string "....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...")

(defn ->board [s]
  (into {}
        (for [[y line] (map-indexed vector (str/split-lines s))
              [x c] (map-indexed vector line)]
          [[x y] c])))

(defn process [s]
  (let [board (->board s)]
    {:walls (into #{}
                  (comp
                    (filter (fn [[_ v]] (= v \#)))
                    (map first))
                  board)
     :start (ffirst (filter (fn [[_ v]] (= v \^)) board))
     :locs (into #{} (map first) board)}))

(def data (process data-string)) 
(def test-data (process test-data-string))

(defn north [[x y]] [x (dec y)])
(defn south [[x y]] [x (inc y)])
(defn east [[x y]] [(inc x) y])
(defn west [[x y]] [(dec x) y])

(def dir-fn
  {:north north
   :south south
   :east east
   :west west})

(def rotate90
  {:north :east
   :east :south
   :south :west
   :west :north})

(defn new-state [data]
  {:loc (data :start) :dir :north :terminated false})

(defn step 
  "Implement a single step of simulation."
  [{walls :walls locs :locs} {:keys [loc dir terminated] :as state}]
  (if terminated
    ; If terminated, do nothing
    state
    (let [new-loc ((dir-fn dir) loc)]
      (cond
        ; if we hit a wall, rotate
        (walls new-loc) (update state :dir rotate90)
        ; if the location is valid, move
        (locs new-loc) (assoc state :loc new-loc)
        ; otherwise mark as terminated
        :else (assoc state :terminated true)))))

(defn simulate [data]
  (take-while 
    (complement :terminated)
    (iterate (partial step data) (new-state data))))

(defn part-1
  "Figure out how many locations are visited."
  [data]
  (transduce
    (comp
      (map :loc)
      (distinct)
      (map (constantly 1)))
    +
    0
    (simulate data)))

(def ans1 (part-1 data))

(test/deftest test-part-1
  (test/is (= 41 (part-1 test-data)))
  (test/is (= 5331 ans1)))

;; ## Part 2


(defn duplicates? [states]
  (= :loop (reduce (fn [acc x] 
                     (if (acc x) 
                       (reduced :loop) 
                       (conj acc x)))
                   #{}
                   states)))

(defn part-2
  "Find all of the locations that create a time loop."
  [data]
  (transduce
    (comp
      (map :loc)
      (distinct)
      (map (partial update-in data [:walls] conj))
      (map simulate)
      (filter duplicates?)
      (map (constantly 1)))
    +
    (simulate data)))

(defonce ans2 (part-2 data))

(test/deftest test-part-2
  (test/is (= 6 (part-2 test-data)))
  (test/is (= 1812 ans2)))


;; #Main

(comment
  (test/run-tests))

(defn -main [& _args]
  (println "Answer1:" ans1)
  (println "Answer2:" ans2))

